File: packages/core/src/tools/pg-query.ts
import { z } from "zod";
import { ActionRegistry, ActionContext, ToolDefinition } from "../types.js";
import { readHandler } from "../actions/query/read.js";
import { writeHandler } from "../actions/query/write.js";
import { explainHandler } from "../actions/query/explain.js";
import { transactionHandler } from "../actions/query/transaction.js";

const queryRegistry: ActionRegistry = {
    read: readHandler,
    write: writeHandler,
    explain: explainHandler,
    transaction: transactionHandler,
};

export const PgQuerySchema = z.object({
    action: z.enum(["read", "write", "explain", "transaction"]).describe("Action to perform: read (SELECT), write (INSERT/UPDATE/DELETE), explain (query plan), transaction (batch)"),
    sql: z.string().optional().describe("SQL query to execute (required for read, write, explain)"),
    params: z.array(z.any()).optional().describe("Query parameters ($1, $2, etc.) for parameterized queries"),
    analyze: z.boolean().optional().describe("Run EXPLAIN ANALYZE (only for 'explain' action)"),
    operations: z.array(z.object({
        sql: z.string(),
        params: z.array(z.any()).optional()
    })).optional().describe("List of operations for transaction batch (only for 'transaction' action)"),
    session_id: z.string().optional().describe("Session ID for transactional context"),
    autocommit: z.boolean().optional().describe("Set to true to allow write/transaction operations without an explicit session"),
});

export async function pgQueryHandler(params: any, context: ActionContext) {
    const handler = queryRegistry[params.action];
    if (!handler) {
        throw new Error(`Unknown action: ${params.action}`);
    }
    return await handler.handler(params, context);
}

export const pgQueryTool: ToolDefinition = {
    name: "pg_query",
    config: {
        description: `Execute SQL queries for data manipulation (DML) and raw read/write operations.

Actions:
  • read: Execute SELECT queries (safe, read-only)
  • write: Execute INSERT/UPDATE/DELETE (requires session_id OR autocommit:true)
  • explain: Analyze query execution plans (supports EXPLAIN ANALYZE)
  • transaction: Execute multiple statements atomically (stateless batch)

Safety: Write operations use Default-Deny policy to prevent accidental data corruption.
Without session_id or autocommit:true, writes will fail with a safety error.

Examples:
  {"action": "read", "sql": "SELECT * FROM users LIMIT 10"}
  {"action": "write", "sql": "UPDATE users SET active = true", "autocommit": true}
  {"action": "transaction", "operations": [{"sql": "..."}]}`,
        inputSchema: PgQuerySchema,
        destructiveHint: true, // Contains write/transaction actions that can modify data
    },
    handler: (context) => (params) => pgQueryHandler(params, context),
};

File: coldquery/tools/pg_query.py
from typing import Any, List, Literal, Optional

from coldquery.actions.query.explain import explain_handler
from coldquery.actions.query.read import read_handler
from coldquery.actions.query.transaction import transaction_handler
from coldquery.actions.query.write import write_handler
from coldquery.core.context import ActionContext
from coldquery.dependencies import CurrentActionContext

# Import the mcp server instance to register the tool
from coldquery.server import mcp

QUERY_ACTIONS = {
    "read": read_handler,
    "write": write_handler,
    "explain": explain_handler,
    "transaction": transaction_handler,
}


@mcp.tool()
async def pg_query(
    action: Literal["read", "write", "explain", "transaction"],
    sql: Optional[str] = None,
    params: Optional[List[Any]] = None,
    analyze: Optional[bool] = None,
    operations: Optional[List[dict]] = None,
    session_id: Optional[str] = None,
    autocommit: Optional[bool] = None,
    context: ActionContext = CurrentActionContext(),
) -> str:
    """Execute SQL queries with safety controls.

    Args:
        action: The type of query action (read, write, explain, transaction)
        sql: SQL query string (required for read, write, explain)
        params: Query parameters for parameterized queries
        analyze: Include ANALYZE in EXPLAIN plans (for explain action)
        operations: List of SQL operations for transaction action
        session_id: Session ID for transactional operations
        autocommit: Enable autocommit for write operations (bypasses session requirement)
        context: ActionContext dependency (injected automatically)

    Returns:
        JSON string containing query results or error information
    """
    handler = QUERY_ACTIONS.get(action)
    if not handler:
        raise ValueError(f"Unknown action: {action}")

    # Prepare params for the handler
    handler_params = {
        "sql": sql,
        "params": params,
        "analyze": analyze,
        "operations": operations,
        "session_id": session_id,
        "autocommit": autocommit,
    }

    return await handler(handler_params, context)
